In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/imdb-review-dataset/part-05.json
/kaggle/input/imdb-review-dataset/part-01.json
/kaggle/input/imdb-review-dataset/part-02.json
/kaggle/input/imdb-review-dataset/part-06.json
/kaggle/input/imdb-review-dataset/sample.json
/kaggle/input/imdb-review-dataset/part-03.json
/kaggle/input/imdb-review-dataset/part-04.json


In [2]:
# importing libraries
import json 
import re 
import pandas as pd

In [3]:
dataDir = "/kaggle/input/imdb-review-dataset/" 
# creating list of files to load
files_to_load = ["part-01.json", "part-02.json", "part-03.json", 
                 "part-04.json", "part-05.json", "part-06.json"]

reviews_json = list() # creating an empty list

In [4]:
# loading data 
for f in files_to_load:
    with open(f"{dataDir}/{f}", mode='r') as file:
        new_reviews = json.load(file)
        for review in new_reviews:
            reviews_json.append(review)

In [5]:
# Converting list of dicts to pandas dataframe
reviews_df = pd.DataFrame(reviews_json)

print(reviews_df.shape) #getting the shape of dataframe

#getting the first 5 rows of the dataframe
reviews_df.head()

(5571499, 9)


review_id         reviewer                          movie rating  \
0  rw5704482    raeldor-96879            After Life (2019– )      9   
1  rw5704483          dosleeb  The Valhalla Murders (2019– )      6   
2  rw5704484  brightconscious           Special OPS (2020– )      7   
3  rw5704485       gasconyway              #BlackAF (2020– )      8   
4  rw5704487     mmason-15867             The Droving (2020)      2   

          review_summary review_date  spoiler_tag  \
0   Very Strong Season 2  3 May 2020            0   
1  Icelandic detectives?  3 May 2020            0   
2        Nothing special  3 May 2020            0   
3               Good but  3 May 2020            0   
4       An honest review  3 May 2020            0   

                                       review_detail   helpful  
0  I enjoyed the first season, but I must say I t...    [1, 1]  
1  I know Iceland is a small country and police d...    [2, 2]  
2  Except K K , no other actor looks comfortable ...    [0, 0]  
3  I'm guessing that as a 62 year old white woman...    [5, 9]  
4  Here's the truth. There's not much to this mov...  [26, 41]

In [6]:
# checking variable types
reviews_df.dtypes

review_id         object
reviewer          object
movie             object
rating            object
review_summary    object
review_date       object
spoiler_tag        int64
review_detail     object
helpful           object
dtype: object

**Which movie had the highest rating ?**

From the value counts below, we can see that the movie 'Avengers: Endgame (2019)' was the most rated. It had the highest number of ratings

In [7]:
# checking the value counts for movie
reviews_df['movie'].value_counts()

Avengers: Endgame (2019)                                           8771
The Shawshank Redemption (1994)                                    8236
Dil Bechara (2020)                                                 7764
Game of Thrones: The Iron Throne (2019) Season 8, Episode 6        7428
Captain Marvel (2019)                                              7158
                                                                   ... 
Confessions of a Teenage Anal Queen (2006 Video)                      1
Lou Grant: Campesinos (1981) Season 4, Episode 16                     1
Lou Grant: Violence (1981) Season 4, Episode 18                       1
Big John, Little John: Off the Wall (1976) Season 1, Episode 10       1
Saibogu 009: Kaijû senso (1967)                                       1
Name: movie, Length: 453528, dtype: int64

**How many reviewers gave it the highest rating?**

The movie had a total of 8771 reviews

**What's its least rating and review detail?**

First we will group the dataframe by the 'movie' column and then create a dataframe for the movie 'Avengers: Endgame (2019)'.

Then from this dataframe, we'll find the least rating given to the movie. (*The least rating given to the movie is 1 and 527 reviewers gave it this*)

Then we'll create a dataframe for just the reviews with the least rating and get their review details.

In [8]:
df = reviews_df
df = df.groupby(['movie'])#grouping by movie

# getting a dataframe with just 'Avengers: Endgame (2019)' movie
endgame_df = df.get_group('Avengers: Endgame (2019)')
endgame_df

review_id            reviewer                     movie rating  \
1061     rw5506515         henrycoles9  Avengers: Endgame (2019)     10   
2104     rw6051980  vittoriocosanostra  Avengers: Endgame (2019)     10   
8174     rw5705866           jejesicaa  Avengers: Endgame (2019)     10   
9938     rw5508642           vishwathg  Avengers: Endgame (2019)     10   
16541    rw5884278        GerryofNorVA  Avengers: Endgame (2019)      6   
...            ...                 ...                       ...    ...   
2992031  rw5068018          erolmehmed  Avengers: Endgame (2019)     10   
2992083  rw5068078          spookybird  Avengers: Endgame (2019)      1   
2992430  rw5068483   fjfairchild-20360  Avengers: Endgame (2019)     10   
2992442  rw5068496   Darkside-Reviewer  Avengers: Endgame (2019)      5   
2992467  rw5068522            Drat1964  Avengers: Endgame (2019)      1   

                                            review_summary       review_date  \
1061     Only the second movie that I chose to watch fo...  24 February 2020   
2104                            Cinema is so beautiful ...  1 September 2020   
8174                    Avengers is a Journey, not movies.        4 May 2020   
9938                            Thw movie of the millenium  25 February 2020   
16541                    An Endgame designed for pre-teens       5 July 2020   
...                                                    ...               ...   
2992031                            You must not miss this!    18 August 2019   
2992083                             Colossal waste of time    18 August 2019   
2992430                      A Climactic End To All Things    19 August 2019   
2992442                                      Total Cop Out    19 August 2019   
2992467                                   Plot holes galor    19 August 2019   

         spoiler_tag                                      review_detail  \
1061               0  I had been out of the loop over Marvel movies ...   
2104               0  Since 2008, films not only from Marvel, but al...   
8174               0  Watched Endgame for the 2nd time, and decided ...   
9938               0  This is the movie of my life and the best eger...   
16541              0  Never has so much been spent on glorious VFX t...   
...              ...                                                ...   
2992031            0  If you are even a remote fan of this universe ...   
2992083            0  Horrible movie. Boring and poorly written. I d...   
2992430            0  Despite the negative responses from people wit...   
2992442            1  I am a big fan of Marvel and the Avengers and ...   
2992467            1  As others have said time travel is always a co...   

            helpful  
1061         [2, 7]  
2104        [6, 14]  
8174        [3, 12]  
9938         [2, 9]  
16541       [2, 14]  
...             ...  
2992031    [38, 61]  
2992083  [176, 308]  
2992430    [39, 63]  
2992442      [4, 8]  
2992467  [247, 434]  

[8771 rows x 9 columns]

In [9]:
# getting value counts for rating
endgame_df['rating'].value_counts()

10    4363
9     1013
8      673
7      604
1      527
6      514
5      368
4      234
3      218
2      159
Name: rating, dtype: int64

From the value counts of the rating we can see that the list rating is 1. So a dataframe will be created for the reviewers who gave the movie a rating of 1, so that we can get their review details from the dataframe.

In [10]:
df = endgame_df
df = df.groupby(['rating'])#grouping by rating

# getting a dataframe with just a rating of 1
low_reviews_df = df.get_group('1')
low_reviews_df

review_id             reviewer                     movie rating  \
161923   rw5727477     kosaildiko-70264  Avengers: Endgame (2019)      1   
178486   rw5904506            kminotaur  Avengers: Endgame (2019)      1   
226689   rw6086020           reffereffe  Avengers: Endgame (2019)      1   
232472   rw5910952  rishabhgautam-24416  Avengers: Endgame (2019)      1   
308977   rw5920766           ctbx-87729  Avengers: Endgame (2019)      1   
...            ...                  ...                       ...    ...   
2985573  rw5064060      mikejenks-50038  Avengers: Endgame (2019)      1   
2988302  rw5066050    luukasleohietanen  Avengers: Endgame (2019)      1   
2988348  rw5066103              pmama69  Avengers: Endgame (2019)      1   
2992083  rw5068078           spookybird  Avengers: Endgame (2019)      1   
2992467  rw5068522             Drat1964  Avengers: Endgame (2019)      1   

                                review_summary        review_date  \
161923             This is totaly stupid movie        11 May 2020   
178486                            Trash movie!       13 July 2020   
226689   Horrible movie, confusing time lines.  11 September 2020   
232472            Ruined my favorite Superhero       16 July 2020   
308977                          Disney Garbage       20 July 2020   
...                                        ...                ...   
2985573                          Waste of time     17 August 2019   
2988302              most overrated movie ever     17 August 2019   
2988348                               Pathetic     17 August 2019   
2992083                 Colossal waste of time     18 August 2019   
2992467                       Plot holes galor     19 August 2019   

         spoiler_tag                                      review_detail  \
161923             0  Oh my God! This movie is horrible... stupid st...   
178486             0  Boring, and too campy. But, hey it's what Marv...   
226689             0  So something else with your time. The movie is...   
232472             0  They ruined hulk very badly, because even Iron...   
308977             0  Marvel, after saturating the market for the pa...   
...              ...                                                ...   
2985573            0  I haven't really been a fan of many of the mar...   
2988302            0  So boring and predictable story with cringe mo...   
2988348            0  Boring. Slow. Badly written, poor story.... wo...   
2992083            0  Horrible movie. Boring and poorly written. I d...   
2992467            1  As others have said time travel is always a co...   

            helpful  
161923     [45, 72]  
178486     [29, 44]  
226689     [32, 49]  
232472     [32, 49]  
308977     [32, 49]  
...             ...  
2985573  [194, 341]  
2988302  [219, 388]  
2988348  [231, 410]  
2992083  [176, 308]  
2992467  [247, 434]  

[527 rows x 9 columns]

In [11]:
# creating  a dataframe with just review details for reviewers with least rating
reviews = low_reviews_df[['review_detail']]
print(reviews) 

                                             review_detail
161923   Oh my God! This movie is horrible... stupid st...
178486   Boring, and too campy. But, hey it's what Marv...
226689   So something else with your time. The movie is...
232472   They ruined hulk very badly, because even Iron...
308977   Marvel, after saturating the market for the pa...
...                                                    ...
2985573  I haven't really been a fan of many of the mar...
2988302  So boring and predictable story with cringe mo...
2988348  Boring. Slow. Badly written, poor story.... wo...
2992083  Horrible movie. Boring and poorly written. I d...
2992467  As others have said time travel is always a co...

[527 rows x 1 columns]
